src/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs

```diff
@@ -205,25 +205,33 @@ private void SetCollectionObjectWithCollectionMember(ref object collection, Coll
                     collection = ReflectionCreateObject(collectionType);
                 }
 
-                if (typeof(IList).IsAssignableFrom(collectionType))
+                AddObjectsIntoTargetCollection(collection, collectionMember, collectionType);
+            }
+        }
+
+        private static void AddObjectsIntoTargetCollection(object targetCollection, List<object> sourceCollection, Type targetCollectionType)
+        {
+            var targetList = targetCollection as IList;
+            if (targetList != null)
+            {
+                foreach (object item in sourceCollection)
                 {
-                    foreach (var item in collectionMember)
-                    {
-                        ((IList)collection).Add(item);
-                    }
+                    targetList.Add(item);
                 }
-                else
+            }
+            else
+            {
+                MethodInfo addMethod = targetCollectionType.GetMethod("Add");
+                if (addMethod == null)
                 {
-                    foreach (var item in collectionMember)
-                    {
-                        MethodInfo addMethod = collectionType.GetMethod("Add");
-                        if (addMethod == null)
-                        {
-                            throw new InvalidOperationException("(addMethod == null)");
-                        }
+                    throw new InvalidOperationException("addMethod == null");
+                }
 
-                        addMethod.Invoke(collection, new object[] { item });
-                    }
+                var arguments = new object[1];
+                foreach (object item in sourceCollection)
+                {
+                    arguments[0] = item;
+                    addMethod.Invoke(targetCollection, arguments);
                 }
```

Comment:
If sourceCollection is large, using addMethod.Invoke will be costly. It might be worth benchmarking roughly how many items need to be added to make it worth creating a temporary delegate.